noir/noir-repo/acvm-repo/acir/src/circuit/opcodes/black_box_function_call.rs

```diff
@@ -4,124 +4,161 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
 // Note: Some functions will not use all of the witness
 // So we need to supply how many bits of the witness is needed
+
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
-pub struct FunctionInput {
-    pub witness: Witness,
+pub enum ConstantOrWitnessEnum<F> {
+    Constant(F),
+    Witness(Witness),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct FunctionInput<F> {
+    pub input: ConstantOrWitnessEnum<F>,
     pub num_bits: u32,
 }
 
+impl<F> FunctionInput<F> {
+    pub fn to_witness(&self) -> Witness {
+        match self.input {
+            ConstantOrWitnessEnum::Constant(_) => unreachable!("ICE - Expected Witness"),
+            ConstantOrWitnessEnum::Witness(witness) => witness,
+        }
+    }
+
+    pub fn num_bits(&self) -> u32 {
+        self.num_bits
+    }
+
+    pub fn from_witness(witness: Witness, num_bits: u32) -> FunctionInput<F> {
+        FunctionInput { input: ConstantOrWitnessEnum::Witness(witness), num_bits }
+    }
+
+    pub fn constant(value: F, num_bits: u32) -> FunctionInput<F> {
+        FunctionInput { input: ConstantOrWitnessEnum::Constant(value), num_bits }
```

Comment:
Can you update the name of one of these to keep them consistent?

Reply:
Done